feat(serve): Add graceful Ctrl+C handling with exit message

afyef opened the pull request:
# Graceful Ctrl+C Handling

Implements reliable Ctrl+C handling for BAML's Python interface with a clean shutdown message. This fixes issues with signal handling across the Python/Rust boundary and improves user experience when stopping BAML processes.

## Changes
- Added custom signal handling using the `ctrlc` crate
- Implemented clean "Shutting Down BAML..." message on interrupt
- Bypassed Python's signal handling to ensure reliable behavior
- Added proper exit code (130) following Unix conventions
- Added detailed documentation explaining the implementation

## Testing
- Tested Python CLI server start/stop
- Manual Ctrl+C testing in various scenarios
- Verified clean shutdown message display
- Tested cross-platform signal handling behavior

## Technical Details
- Uses `ctrlc` crate for reliable cross-platform signal handling
- Implements channel-based communication between signal handler and main thread
- Follows Unix conventions for interrupt handling (exit code 130)
- Properly documented implementation rationale and technical decisions

## Notes
- While this is a workaround, it provides reliable interrupt handling without requiring major architectural changes
- The implementation bypasses Python's signal handling entirely to avoid conflicts
- Future improvements could involve deeper integration with BAML's architecture

<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Enhances BAML's Python interface with reliable Ctrl+C handling and clean shutdown message using the `ctrlc` crate.
> 
>   - **Behavior**:
>     - Implements custom signal handling to bypass Python's signal handlers
>     - Displays "Shutting Down BAML..." message on interrupt
>     - Ensures proper process termination with standard exit codes
>   - **Implementation**:
>     - Uses `ctrlc` crate for cross-platform signal handling
>     - Implements channel-based communication for safe signal handling
>     - Follows Unix conventions with exit code 130 for SIGINT
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup> for the latest commit. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

sxlijin commented:
Started a discussion thread in https://discord.com/channels/1119368998161752075/1316923453814734959 - I have some questions about how this plays with pyo3 and different architectures (it sounds like we handle this fine in N-API world?)

Thanks for digging into this!

sxlijin commented on engine/baml-runtime/src/cli/serve/mod.rs:
```diff
@@ -340,7 +330,29 @@ Tip: test that the server is up using `curl http://localhost:{}/_debug/ping`
+            }
+        };
+
+        // Set ctrl-c handler to ignore so Python doesn't intercept it, otherwise python doesnt like it
```
As discussed in Discord, let's try moving this into [`invoke_runtime_cli`](https://github.com/BoundaryML/baml/blob/canary/engine/language_client_python/src/lib.rs#L12-L13)

Everything in `baml-runtime` is shared across all 4 builds of BAML (Python, Node, Ruby, and the internal-only Rust CLI) so I'd rather not muck with default signal handling behavior unless we have to - e.g. right now it's really easy for someone to misunderstand how the current code works, because it's easy to read it as setting up a tokio ctrl-c handler and then uninstalling all signal handlers (even though it's actually the other way around due to how Rust async works)

sxlijin commented on engine/baml-runtime/src/cli/serve/mod.rs:
```diff
@@ -313,21 +317,7 @@ impl Server {
             tcp_listener,
             app.layer(axum::middleware::from_fn(Server::auth_middleware)),
         );
-        // TODO: we do not handle this ourselves, because tokio's default
-        // handling is pretty good on unix.
-        //
-        // Not totally sure if this WAI on
-        // windows, but there are some potential pitfalls that we can run into
-        // if we try to handle this ourselves, see
-        // https://docs.rs/tokio/latest/tokio/signal/fn.ctrl_c.html#caveats
-        //
-        // Namely- we need to ensure resilient delivery of Ctrl-C to everything, and I
-        // suspect we need to do a bit of work to ensure that we handle that bookkeeping
-        // correctly. Shutting down the BAML runtime, tokio runtime, _and_ axum is not
-        // super straightforward, because I don't know how much is handled for us
-        // out of the box.
-        //
-        // .with_graceful_shutdown(signal::ctrl_c());
```
Can we keep these comments around in this file? Even if we push the libc signal handler stuff to language_client_python/src/lib.rs. Mostly just because this is because a good reminder for myself about how careful we should be about touching signal handling code here.

afyef committed:
feat(python): add ctrlc dependency for signal handling

afyef committed:
feat(python): implement graceful ctrl+c handling

This implements a custom signal handling mechanism using the ctrlc crate to: 1. Bypass Python's signal handling 2. Provide consistent behavior across platforms 3. Ensure graceful shutdown with proper exit codes

While this is a workaround, it provides reliable interrupt handling without requiring major architectural changes to BAML's signal handling.

afyef committed:
Merge branch 'canary' into feat/graceful-ctrl-c

sxlijin reviewed (approved):
This is clever! Nice work with this

We could use a oneshot channel for the signal but I really like this - on the one hand spawning an additional thread feels overkill, but it's also a neat way to avoid wiring in `Python::check_signals` everywhere

afyef committed:
Merge branch 'canary' into feat/graceful-ctrl-c

sxlijin commented:
Sorry for the delay! I was in a GitHub beta for a new UI and couldn't figure out how to trigger workflows on your PR (maintainers need to explicitly approve PRs from new contributors)

sxlijin commented:
Hmm - looks like I can't push to your branch to update `Cargo.lock`. Can you commit that change?

afyef committed:
chore: update Cargo.lock

afyef commented:
@sxlijin  Committed.
